seqlock/seqlock.c

```diff
@@ -1,50 +1,46 @@
 #include <assert.h>
+#include <stdatomic.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 
 #include "seqlock.h"
 
 #define SEQLOCK_WRITER 1U
 
-#if defined(__i386__) || defined(__x86_64__)
-#define spin_wait() __builtin_ia32_pause()
-#elif defined(__aarch64__)
-#define spin_wait() __asm__ __volatile__("isb\n")
+#if defined(__i386__) || defined(__x86_64__) || defined(__aarch64__)
+#define spin_wait() atomic_thread_fence(memory_order_seq_cst)
```

Comment:
You have to confirm the generated code is identical to the intended instruction sequence.

Reply:
It would be generating 
```
dmb     ish
```
with Apple clang version 14.0.3 (clang-1403.0.22.14.1).

I would figure what's the output in gcc x86

Reply:
In gcc 13.2.1 with x86-64, both
1. using __builtin_ia32_pause()
2. using atomic_thread_fence(memory_order_seq_cst)
are generating 
```
data16 cs nopw 0x0(%rax,%rax,1)
nop
xchg   %ax,%ax
```

Reply:
In LLVM clang version 15.0.7 with x86-64, adding an extra `pause` when __builtin_ia32_pause is used.

```
    125 0000000000001190 <seqlock_acquire_rd>:
    126     1190:       50                      push   %rax
    127     1191:       0f ae f0                mfence
    128     1194:       8b 07                   mov    (%rdi),%eax
    129     1196:       a8 01                   test   $0x1,%al
    130     1198:       75 08                   jne    11a2 <seqlock_acquire_rd+0x12>
    131     119a:       a8 01                   test   $0x1,%al
    132     119c:       75 0f                   jne    11ad <seqlock_acquire_rd+0x1d>
    133     119e:       59                      pop    %rcx
    134     119f:       c3                      ret
    135     11a0:       f3 90                   pause
    136     11a2:       0f ae f0                mfence
    137     11a5:       8b 07                   mov    (%rdi),%eax
    138     11a7:       a8 01                   test   $0x1,%al
    139     11a9:       75 f5                   jne    11a0 <seqlock_acquire_rd+0x10>
    140     11ab:       eb ed                   jmp    119a <seqlock_acquire_rd+0xa>
    141     11ad:       48 8d 3d 6f 0e 00 00    lea    0xe6f(%rip),%rdi        # 2023 <_IO_stdin_used+0x23>
    142     11b4:       48 8d 35 82 0e 00 00    lea    0xe82(%rip),%rsi        # 203d <_IO_stdin_used+0x3d>
    143     11bb:       48 8d 0d 85 0e 00 00    lea    0xe85(%rip),%rcx        # 2047 <_IO_stdin_used+0x47>
    144     11c2:       ba 40 00 00 00          mov    $0x40,%edx
    145     11c7:       e8 84 fe ff ff          call   1050 <__assert_fail@plt>
    146     11cc:       0f 1f 40 00             nopl   0x0(%rax)
```